### DISCUSSION TITLE
Compiling on M1 Mac

### DISCUSSION THREAD

**Discussion by maxnezdyur** [KratosMultiphysics/Kratos#9837] [General]
Has anyone had experience compiling on an M1 Mac?
I am having the problem below 

> ld: warning: direct access in function 'void Kratos::Serializer::Register<Kratos::ShellThickElement3D4N<(Kratos::ShellKinematics)0> >(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Kratos::ShellThickElement3D4N<(Kratos::ShellKinematics)0> const&)' from file 'CMakeFiles/KratosStructuralMechanicsCore.dir/structural_mechanics_application.cpp.o' to global weak symbol 'typeinfo name for Kratos::ShellThickElement3D4N<(Kratos::ShellKinematics)0>' from file 'CMakeFiles/KratosStructuralMechanicsCore.dir/custom_elements/shell_thick_element_3D4N.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
> 
I tried to change boosts to -fvisibility=hidden but to no avail. I was wondering if someone knew a fix. 
Best regards.

**Comment by ghost** [comment]
I also tried to compile on my M1 machine and failed. I think the problem is not those warnings but linker problem; undefined symbol(s) for architecture arm64.

**Reply by loumalouomega** [reply]
Huuum, we can try to debug it with a RPi in 64 bits mode...

**Reply by loumalouomega** [reply]
I just did a compilation with an Ampere server (ARM64, see https://levelup.gitconnected.com/a-powerful-server-from-oracle-cloud-always-free-cbc73d9fbfee). And the compilation works fine. I think the problem is coming from the Mac OS linker (my experience with the MacOS linker is not very positive). Have you tried to use the GCC and a linker from Homebrew?

**Reply by loumalouomega** [reply]
I added an architecture detection: https://github.com/KratosMultiphysics/Kratos/pull/10098

And as you can see it works in ARM64:

![image](https://user-images.githubusercontent.com/19991680/182575622-14a6cd15-d2a7-4290-8ee6-e66d5c09a9d5.png)

**Reply by loumalouomega** [reply]
Using LD as linker and gcc as compiler

**Reply by ghost** [reply]
Hi @loumalouomega ,

Thank you for your response. This looks promising, can you share a snip of your configure.sh file as a reference for us?

**Reply by loumalouomega** [reply]
It is almost the default, the machine uses Oracle OS, which is a modified version of RedHat:

~~~sh
#!/bin/bash
# Please do not modify this script

# You can use your interpreter of choice (bash, sh, zsh, ...)

# For any question please contact with us in:
#   - https://github.com/KratosMultiphysics/Kratos

# Optional parameters:
# You can find a list will all the compiation options in INSTALL.md or here:
#   - https://github.com/KratosMultiphysics/Kratos/wiki/Compilation-options

# Function to add apps
add_app () {
    export KRATOS_APPLICATIONS="${KRATOS_APPLICATIONS}$1;"
}

# Set compiler
export CC=gcc
export CXX=g++

# Set variables
export KRATOS_SOURCE="${KRATOS_SOURCE:-"$( cd "$(dirname "$0")" ; pwd -P )"/..}"
export KRATOS_BUILD="${KRATOS_SOURCE}/build"
export KRATOS_APP_DIR="${KRATOS_SOURCE}/applications"
# export KRATOS_INSTALL_PYTHON_USING_LINKS=ON

# Set basic configuration
export KRATOS_BUILD_TYPE=${KRATOS_BUILD_TYPE:-"Release"}
export PYTHON_EXECUTABLE=${PYTHON_EXECUTABLE:-"/usr/bin/python3"}

# Set applications to compile
export KRATOS_APPLICATIONS=
add_app ${KRATOS_APP_DIR}/LinearSolversApplication
add_app ${KRATOS_APP_DIR}/StructuralMechanicsApplication
add_app ${KRATOS_APP_DIR}/FluidDynamicsApplication

# Clean
clear
rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/cmake_install.cmake"
rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/CMakeCache.txt"
rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/CMakeFiles"

# Configure
cmake                                     \
-H"${KRATOS_SOURCE}"                      \
-B"${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}"  \
-DUSE_MPI=OFF                             \
-DUSE_EIGEN_MKL=OFF

# Buid
cmake --build "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}" --target install -- -j$(nproc)
~~~

**Reply by loumalouomega** [reply]
As I said, I would try gcc instead of Mac Clang: https://codeforces.com/blog/entry/88063

My experience is that since Apple Silicon, Apple has been changing a lot the system libraries and the way the linker works and it is very troublesome.

**Reply by ghost** [reply]
thank you for the help!

I still have difficulties with my GCC installation. However, I managed to make clang works. Now I just have the following error when compiling the FluidDynamicsApplication but other than that everything seems ok.


<img width="1145" alt="Screenshot 2022-08-03 at 14 16 27" src="https://user-images.githubusercontent.com/108530073/182605441-64ec32d5-ef2f-4b7b-a55c-6d56249a96e6.png">

**Reply by ghost** [reply]
Here is an example of the test I ran for LinearSolver

<img width="1147" alt="Screenshot 2022-08-03 at 14 18 28" src="https://user-images.githubusercontent.com/108530073/182605907-85050026-f6d8-4868-bcde-9c55acc82e09.png">

**Reply by ghost** [reply]
Here is my configuration.sh if anyone needs


```
#!/bin/bash

# Please do not modify this script

# You can use your interpreter of choice (bash, sh, zsh, ...)

# For any question please contact with us in:
#   - https://github.com/KratosMultiphysics/Kratos

# Optional parameters:
# You can find a list will all the compiation options in INSTALL.md or here:
#   - https://github.com/KratosMultiphysics/Kratos/wiki/Compilation-options

# Function to add apps
add_app () {
    export KRATOS_APPLICATIONS="${KRATOS_APPLICATIONS}$1;"
}

# Set compiler
export CC=/opt/homebrew/opt/llvm/bin/clang
export CXX=/opt/homebrew/opt/llvm/bin/clang++

# Set variables
export KRATOS_SOURCE="${KRATOS_SOURCE:-"$( cd "$(dirname "$0")" ; pwd -P )"/..}"
export KRATOS_BUILD="${KRATOS_SOURCE}/build"
export KRATOS_APP_DIR="${KRATOS_SOURCE}/applications"
export KRATOS_INSTALL_PYTHON_USING_LINKS=ON

# Set basic configuration
export KRATOS_BUILD_TYPE=${KRATOS_BUILD_TYPE:-"Release"}
export BOOST_ROOT=${BOOST_ROOT:-"/Users/rezhadrian/Projects/boost_1_79_0"}
export PYTHON_EXECUTABLE=${PYTHON_EXECUTABLE:-"/opt/anaconda3/bin/python3"}

# Set applications to compile
export KRATOS_APPLICATIONS=
add_app ${KRATOS_APP_DIR}/LinearSolversApplication
add_app ${KRATOS_APP_DIR}/StructuralMechanicsApplication
add_app ${KRATOS_APP_DIR}/FSIApplication
add_app ${KRATOS_APP_DIR}/FluidDynamicsApplication

# Clean
clear
rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/cmake_install.cmake"
rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/CMakeCache.txt"
rm -rf "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}/CMakeFiles"

# Configure
/Applications/CMake.app/Contents/bin/cmake -H"${KRATOS_SOURCE}" -B"${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}" \
 -DCMAKE_CXX_FLAGS="${CMAKE_CXX_FLAGS} -msse3 -std=c++11 -Qunused-arguments -Wno-undefined-var-template -Wno-expansion-to-defined -L/opt/homebrew/opt/llvm/lib" \
 -DCMAKE_C_FLAGS="${CMAKE_C_FLAGS} -msse3 -Qunused-arguments -Wno-undefined-var-template -Wno-expansion-to-defined -L/opt/homebrew/opt/llvm/lib" \
 -DUSE_EIGEN_MKL=OFF

# Buid
/Applications/CMake.app/Contents/bin/cmake --build "${KRATOS_BUILD}/${KRATOS_BUILD_TYPE}" --target install -- -j$(sysctl -n hw.physicalcpu)

**Reply by ghost** [reply]
Using LLVM from Brew instead of what's included with XCode

**Reply by loumalouomega** [reply]
> Here is an example of the test I ran for LinearSolver
> 
> <img alt="Screenshot 2022-08-03 at 14 18 28" width="1147" src="https://user-images.githubusercontent.com/108530073/182605907-85050026-f6d8-4868-bcde-9c55acc82e09.png">

Can yu try to compile without MKL?, it is a x86_64 library.

**Reply by loumalouomega** [reply]
Okay I see that is not included, in that case that is right, you are not supposed to be able to use MKL solvers in a ARM64 architecture.

**Reply by loumalouomega** [reply]
> thank you for the help!
> 
> I still have difficulties with my GCC installation. However, I managed to make clang works. Now I just have the following error when compiling the FluidDynamicsApplication but other than that everything seems ok.
> 
> <img alt="Screenshot 2022-08-03 at 14 16 27" width="1145" src="https://user-images.githubusercontent.com/108530073/182605441-64ec32d5-ef2f-4b7b-a55c-6d56249a96e6.png">

That error looks like something pingable to @rubenzorrilla

**Reply by rubenzorrilla** [reply]
This is @mrhashemi's code. Not mines.

**Reply by loumalouomega** [reply]
> This is @mrhashemi's code. Not mines.

But at least you know who
